flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushPartitionIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,320 @@
+import org.apache.flink.table.planner.plan.stats.FlinkStatistic;
+import org.apache.flink.table.planner.plan.utils.FlinkRelOptUtil;
+import org.apache.flink.table.planner.plan.utils.PartitionPruner;
+import org.apache.flink.table.planner.plan.utils.RexNodeExtractor;
+import org.apache.flink.table.planner.plan.utils.RexNodeToExpressionConverter;
+import org.apache.flink.table.planner.utils.CatalogTableStatisticsConverter;
+import org.apache.flink.table.types.logical.LogicalType;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.core.Filter;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexShuttle;
+import org.apache.calcite.rex.RexUtil;
+import org.apache.calcite.tools.RelBuilder;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TimeZone;
+import java.util.stream.Collectors;
+
+import scala.Option;
+import scala.Tuple2;
+import scala.collection.JavaConversions;
+import scala.collection.Seq;
+
+/**
+ * Planner rule that tries to push partition evaluated by filter condition into a {@link LogicalTableScan}.
+*/
+public class PushPartitionIntoTableSourceScanRule extends RelOptRule {
+	public static final PushPartitionIntoTableSourceScanRule INSTANCE = new PushPartitionIntoTableSourceScanRule();
+
+	public PushPartitionIntoTableSourceScanRule(){
+		super(operand(Filter.class,
+				operand(LogicalTableScan.class, none())),
+			"PushPartitionTableSourceScanRule");
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		Filter filter = call.rel(0);
+		if (filter.getCondition() == null) {
+			return false;
+		}
+		TableSourceTable tableSourceTable = call.rel(1).getTable().unwrap(TableSourceTable.class);
+		if (tableSourceTable == null){
+			return false;
+		}
+		DynamicTableSource dynamicTableSource = tableSourceTable.tableSource();
+		if (!(dynamicTableSource instanceof SupportsPartitionPushDown)) {
+			return false;
+		}
+		CatalogTable catalogTable = tableSourceTable.catalogTable();
+		if (!catalogTable.isPartitioned() || catalogTable.getPartitionKeys().size() == 0) {
+			return false;
+		}
+		return Arrays.stream(tableSourceTable.extraDigests()).noneMatch(digest -> digest.startsWith("partitions=["));
+	}
+
+	@Override
+	public void onMatch(RelOptRuleCall call) {
+		Filter filter = call.rel(0);
+		LogicalTableScan scan = call.rel(1);
+		FlinkContext context = call.getPlanner().getContext().unwrap(FlinkContext.class);
+		TableSourceTable tableSourceTable = scan.getTable().unwrap(TableSourceTable.class);
+		// use new dynamic table source to push down
+		DynamicTableSource dynamicTableSource = tableSourceTable.tableSource().copy();
```

Comment:
Initialize variables as close as possible to the point where you use them, please refer to [The Principle of Proximity](https://www.approxion.com/the-principle-of-proximity/) for more detailes